File: Coursera/Algorithms_Part_1/Ex6BinarySearch.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;

import java.util.Arrays;

public class Ex6BinarySearch {

    private static int find(int[] a, int i) {
        Arrays.sort(a);
        int lo = 0;
        int hi = a.length;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (i < a[mid]) hi = mid - 1;
            else if (i > a[mid]) lo = mid + 1;
            else return mid;
        }
        return -1;
    }

    public static void main(String[] args) {
        int n = 100;
        int[] value = new int[n];
        StdOut.print("[");
        for (int i = 1; i <= n; i++) {
            value[i - 1] = StdRandom.uniform(-100, 100);
            StdOut.print(value[i - 1]);
            if (i < n) {
                StdOut.print(",");
                if (i % 50 == 0) StdOut.println();
            }
        }
        StdOut.println("]");

        int testCases = StdRandom.uniform(n / 4);
        for (int i = 0; i < testCases; i++) {
            int test = StdRandom.uniform(-100, 100);
            int index = Ex6BinarySearch.find(value, test);
            StdOut.println("Find " + test + ":" + (index == -1 ? " Not Found" : index));
        }
    }
}

File: PythonLearnig/Algorithms_Part_1/Ex6BinarySearch.py
import math


def find(sorted_list, value):

    lo, hi = 0, len(sorted_list) - 1

    while lo <= hi:
        mid = lo + int(math.floor((hi - lo) / 2))

        if sorted_list[mid] < value:
            lo = mid + 1
        elif sorted_list[mid] > value:
            hi = mid - 1
        else:
            return mid

    return -1


def test_binary_search():

    import random
    import math
    n = 100
    value = [random.randint(-100, 100) for _ in range(n)]
    value.sort()
    print(value)

    test_cases = [random.randint(-100, 100) for _ in range(math.floor(n / 4))]
    for test in test_cases:
        index = find(value, test)
        if index == -1:
            print(f"Find {test}: Not Found")
        else:
            print(f"Find {test}: {index}")


if __name__ == '__main__':
    test_binary_search()
